File: app/routers/meta.py
"""Meta / read-only endpoints for the Mock Jira Server (Server flavor).

Routes here are mounted under the /rest/api/2 prefix in main.py:
    GET /serverInfo, GET /myself, GET /priority, GET /field,
    GET /status, GET /issuetype, GET /issue/{key}/changelog
"""

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app import shapes
from app.store import get_db

router = APIRouter()


@router.get("/serverInfo")
def server_info():
    return JSONResponse(content=shapes.server_info())


@router.get("/myself")
def myself(request: Request):
    user = getattr(request.state, "user", None)
    if user is None:
        db = get_db()
        if db is not None:
            users = db.list_users()
            if users:
                user = users[0]
    return JSONResponse(content=shapes.myself_shape(user))


@router.get("/priority")
def priority():
    return JSONResponse(content=shapes.priority_list())


@router.get("/field")
def field():
    return JSONResponse(content=shapes.field_list())


@router.get("/status")
def status():
    return JSONResponse(content=shapes.status_list())


@router.get("/issuetype")
def issuetype():
    return JSONResponse(content=shapes.issuetype_list())


@router.get("/issue/{key}/changelog")
def issue_changelog(key: str):
    """Minimal changelog stub. This mock does not track field history, so an
    empty ``histories`` array is the correct answer for any existing issue."""
    db = get_db()
    from app.routers.issues import _resolve

    issue = _resolve(db, key) if db else None
    if issue is None:
        return JSONResponse(
            status_code=404,
            content=shapes.error_body("Issue Does Not Exist"),
        )
    return JSONResponse(
        content={
            "startAt": 0,
            "maxResults": 50,
            "total": 0,
            "histories": [],
        }
    )

File: src/routes/meta.ts
/**
 * Meta / read-only endpoints for the Mock Jira Server (Server flavor).
 *
 * Routes here are mounted under the /rest/api/2 prefix in app.ts:
 *     GET /serverInfo, GET /myself, GET /priority, GET /field,
 *     GET /status, GET /issuetype, GET /issue/:key/changelog
 */

import { FastifyInstance } from "fastify";

import * as shapes from "../shapes";
import { getDb } from "../store";
import { resolveIssue } from "./issues";

export async function metaRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.get("/serverInfo", async (_request, reply) => {
    return reply.send(shapes.serverInfo());
  });

  fastify.get("/myself", async (request, reply) => {
    let user = (request as any).authUser ?? null;
    if (user == null) {
      const db = getDb();
      if (db) {
        const users = await db.listUsers();
        if (users.length) {
          user = users[0];
        }
      }
    }
    return reply.send(shapes.myselfShape(user));
  });

  fastify.get("/priority", async (_request, reply) => {
    return reply.send(shapes.priorityList());
  });

  fastify.get("/field", async (_request, reply) => {
    return reply.send(shapes.fieldList());
  });

  fastify.get("/status", async (_request, reply) => {
    return reply.send(shapes.statusList());
  });

  fastify.get("/issuetype", async (_request, reply) => {
    return reply.send(shapes.issuetypeList());
  });

  // Minimal changelog stub. This mock does not track field history, so an
  // empty `histories` array is the correct answer for any existing issue.
  fastify.get("/issue/:key/changelog", async (request, reply) => {
    const db = getDb();
    const { key } = request.params as { key: string };
    const issue = db ? await resolveIssue(db, key) : null;
    if (issue == null) {
      return reply.code(404).send(shapes.errorBody("Issue Does Not Exist"));
    }
    return reply.send({
      startAt: 0,
      maxResults: 50,
      total: 0,
      histories: [],
    });
  });
}
